### ISSUE TITLE
How many id tokens are used in training

### ISSUE THREAD

**Issue by 2hiTee** [wgsxm/PartCrafter#2]
Wonderful work! I am wondering the number of id tokens used in training process.

**Comment by 2hiTee** [comment]
> We use ID embeddings, which are directly added to other tokens, instead of ID tokens, which may mean token-wise concatenation.
> 
> ID embeddings are flexible during training and determined by the number of parts in the training samples. During inference, users can customize the desired number of parts as shown in Figure 6. Specifically, ID embeddings are implemented via `nn.Embedding` with a max number of parts set to `32`.

Thanks for your reply! So during inference, if I specify N=K, it will chose the first K id tokens to concate?

**Comment by chenguolin** [comment]
Not concate. It will choose the first `K` embeddings and **add** these embeddings to the corresponding latent tokens (`K*N` tokens in total and `N` tokens for each part) of each part.

**Comment by 2hiTee** [comment]
> Not concate. It will choose the first `K` embeddings and **add** these embeddings to the corresponding latent tokens (`K*N` tokens in total and `N` tokens for each part) of each part.

Thanks for your reply again! So in order **not** to let each id token learn specific semantic part, you shuffle the order of the parts during training? Thanks again!

**Comment by 2hiTee** [comment]
> Yes. Please refer to Section 3.2.
> 
> <img alt="Image" width="736" src="https://private-user-images.githubusercontent.com/47670737/453795319-6caece96-289e-4b86-96c9-f919ba08ea85.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk2Mjc3ODksIm5iZiI6MTc0OTYyNzQ4OSwicGF0aCI6Ii80NzY3MDczNy80NTM3OTUzMTktNmNhZWNlOTYtMjg5ZS00Yjg2LTk2YzktZjkxOWJhMDhlYTg1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjExVDA3MzgwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzZWY4NjA1NWU1ODE5ZmUzZjkwY2M5NmZhNWM2YWQ5MzczNTMxMjJjYzM2NzA5M2NhYTA1M2IyMTczYjhlMTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.w6LwWM0dO2tPvvd-2pkiSLL1ngmMayBXt4PvaxAK3rU">

Thanks for your wonderful work! And thanks for your kind reply!
